### PR TITLE
Test out /beta_features on staging

### PR DESCRIPTION
**Pull request by clekstro**
Currently getting a 404 (at least when I try locally).

EDIT: Found the issue. Forgot we're now scoping the `/beta_features` to the individual user.

### CODE REVIEW

**Comment by clekstro** [comment]
So with the changed URL, it looks like I need to ensure we send a `PATCH` instead of a `PUT` which appears to be what we'd send via Ember-Data by default.

**Comment by clekstro** [comment]
I need to do some more fiddling with a `PATCH`, because it's only expecting a sliver of the actual record to be sent across the wire. When I send a `PATCH` with the full record, I get another `id` back as if it created a new record (from 1 to 3). Maybe a bug?

**Comment by clekstro** [comment]
Strangely enough, when I refresh the page, it looks like the feature has been enabled (based on the response and the UI), but still not actually doing anything, so going to dive in and see what's up.

**Comment by clekstro** [comment]
I think I'm making the correct request, but it looks like the id's that I get back aren't valid.

Here's the response I get when fetching the features:

```json
{
  "@type": "beta_features",
  "@href": "/user/4281/beta_features",
  "@representation": "standard",
  "beta_features": [
    {
      "@type": "user_beta_feature",
      "@representation": "standard",
      "id": 3,
      "name": "pride",
      "description": "show your travis pride!",
      "enabled": true,
      "feedback_url": "https://github.com/travis-ci/beta-features/issues/new"
    },
    {
      "@type": "user_beta_feature",
      "@representation": "standard",
      "id": 4,
      "name": "dashboard",
      "description": "Turn on and off our new dashboard layout",
      "enabled": true,
      "feedback_url": "https://github.com/travis-ci/beta-features/issues/new"
    }
  ]
}
```

Sending this request then returns a 404:

```sh
curl 'https://api-staging.travis-ci.org/user/:user_id/beta_feature/4' -X PATCH -H 'Pragma: no-cache' -H 'Origin: http://localhost:4200' -H 'Accept-Encoding: gzip, deflate, sdch, br' -H 'Accept-Language: en-US,en;q=0.8,de;q=0.6' -H 'Travis-API-Version: 3' -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/55.0.2883.75 Safari/537.36' -H 'Content-Type: application/json' -H 'Accept: application/json' -H 'Cache-Control: no-cache' -H 'Authorization: token <my_token>' -H 'Connection: keep-alive' -H 'Referer: http://localhost:4200/features' -H 'DNT: 1' --data-binary '{"beta_feature.enabled":false}' --compressed
```

File: app/adapters/beta-feature.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  auth: service(),
+
+  buildURL(modelName, id, snapshot, requestType) {
+    let url = this._super(...arguments);
+    if (requestType === 'GET') {
+      let replacement = `user/${this.get('auth.currentUser.id')}/beta_features`;
+      url = url.replace(/beta_features/, replacement);
+    } else {
+      if (requestType === 'PUT') { requestType = 'PATCH'; }
+      let replacement = `user/${this.get('auth.currentUser.id')}/beta_feature`;
```
Comment (renee-travisci) [inline app/adapters/beta-feature.js]: I think you need the beta_feature id as part of the URL /user/id/beta_feature/id

**Comment by joecorcoran** [comment]
I don't know what we were doing wrong earlier, but this works for me:

```
curl https://api-staging.travis-ci.org/v3/user/4289/beta_features' -H 'Authorization: token abc'
```

```
curl https://api-staging.travis-ci.org/v3/user/4289/beta_feature/1 -X PATCH -H 'Authorization: token abc'
```

The available beta features on staging have id 1 and 2...

File: app/adapters/beta-feature.js (hunk 1 above)
Comment (clekstro) [inline app/adapters/beta-feature.js]: Right. I'm sending `id`, it's just not apparent from this code as it's automatically appended.

**Comment by joecorcoran** [comment]
Okay, I see the same issue. Once a user has "switched on" a feature, we return the wrong ids in the `/beta_features` endpoint, somehow. Note that this should show `1` and `2` @renee-travisci 

```
curl https://api-staging.travis-ci.org/v3/user/4289/beta_features -H 'Authorization: token sPj_527PhgfTV7tZIBlcKw'
{
  "@type": "beta_features",
  "@href": "/v3/user/4289/beta_features",
  "@representation": "standard",
  "beta_features": [
    {
      "@type": "beta_feature",
      "@representation": "standard",
      "id": 2,
      "name": "dashboard",
      "description": "Turn on and off our new dashboard layout",
      "enabled": false,
      "feedback_url": "https://github.com/travis-ci/beta-features/issues/new"
    },
    {
      "@type": "user_beta_feature",
      "@representation": "standard",
      "id": 5,
      "name": "pride",
      "description": "show your travis pride!",
      "enabled": false,
      "feedback_url": "https://github.com/travis-ci/beta-features/issues/new"
    }
  ]
}
```

**Comment by renee-travisci** [comment]
@joecorcoran @clekstro The GET on /beta_features is now returning the correct id. The id when the object is updated/created via PATCH or DELETE is not the beta_feature_id but the user_beta_feature_id (the model we use in the backend to save the user's settings) and I'm not sure how to get that one to show up correctly yet.

**Comment by clekstro** [comment]
@renee-travisci I've fixed the failing tests in `travis-web`, meaning we now have a PR deployment in the links below. That means you can quickly test any changes you deploy to the staging API (.org anyway) without having to boot things up locally :)

**Comment by clekstro** [comment]
I approved the diffs on [Percy](https://percy.io/travis-ci/travis-web/builds/138982), as they all appear to be the product of the new dashboard/centered design, which is expected on this branch.

**Comment by clekstro** [comment]
@renee-travisci I tried the most recent version on staging, but am still having issues toggling features and getting a `404`. Is it possible just my feature records are affected at the db level?

**Comment by backspace** [comment]
Hmm, it looks to me like the visual diffs show a problem:


This one has no organisation name:

No organisation name diff

No organisation name screenshot

This one has no organisation nor repository name:

No organisation nor repository diff

No organisation nor repository screenshot



Maybe this doesn’t really matter as the headers look fine in the PR deployments, but it seems to me that we should understand why the slug isn’t rendering properly in the acceptance tests, because it probably indicates a divergence between the true and mock APIs.

**Comment by renee-travisci** [comment]
@clekstro the errors you were getting were due to the id issues in the returned result from the PATCH route. I've fixed that on staging and now the beta_features toggles are working as expected here: https://ce-beta-features-endpoint-staging-staging.test-deployments.travis-ci.org/features w00t!

**Comment by lislis** [comment]
Closing because part of https://github.com/travis-ci/travis-web/pull/937

**Pull request closed without merging**